src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "Utilities/TypeTraits/GetFundamentalType.hpp"
+
+namespace PUP {
+class er;
+}  // namespace PUP
+
+/**
+ * \ingroup DataStructuresGroup
+ *
+ * \brief A dynamically sized vector of `DataVector`s. For convenience it can
+ * also be instantiated for fundamental types.
+ *
+ * \details This class is useful when one wants to create a `std::vector<T>`
+ * with a size that is unknown at compile time. It allocates all `DataVector`s
+ * in a single memory chunk rather than allocating each individually. If the
+ * size of the vector is known at compile time, a `TempBuffer` object should be
+ * used instead.
+ *
+ * If needed it should be fairly straightforward to generalize to
+ * `ComplexDataVector`.
+ */
+template <typename T>
+class DynamicBuffer {
+ public:
+  static constexpr bool is_data_vector_type =
+      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T>, T>;
+
+  DynamicBuffer() = default;
+
+  /*!
+   * Constructs a `DynamicBuffer`. The `number_of_vectors` corresponds to the
+   * number of `DataVector`s which are saved inside, each of which has size
+   * `number_of_grid_points`. `number_of_grid_points` has to be 1 if T is a
+   * fundamental type.
+   */
+  DynamicBuffer(size_t number_of_vectors,
+                size_t number_of_grid_points) noexcept;
+  ~DynamicBuffer() = default;
+  DynamicBuffer(DynamicBuffer&& other) noexcept = default;
+  DynamicBuffer& operator=(DynamicBuffer&& other) = default;
+
+  DynamicBuffer(const DynamicBuffer& other) noexcept;
+
+  DynamicBuffer& operator=(const DynamicBuffer& other) noexcept;
+
+  T& operator[](size_t index) { return data_[index]; }
+  T& at(size_t index) { return data_.at(index); }
+  const T& operator[](size_t index) const { return data_[index]; }
+  const T& at(size_t index) const { return data_.at(index); }
+
+  auto begin() { return data_.begin(); }
+  auto end() { return data_.end(); }
+  auto begin() const { return data_.begin(); }
+  auto end() const { return data_.end(); }
+
+  size_t size() const noexcept { return data_.size(); }
+
+  /// \cond
+  // clang-tidy: runtime-references
+  void pup(PUP::er& p) noexcept;
+  /// \endcond
+
+ private:
+  // sets data references for all `data_` into `buffer_`
+  void set_references();
+
+  template <typename LocalT>
+  friend bool operator==(const DynamicBuffer<LocalT>& lhs,
```

Comment:
I think you need to `// NOLINT` this line because clang-tidy has a bug in it (or at least as far as I can tell it's a bug)

Reply:
squashed, thank you for your review :)